File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)

        if n < 3: 
            return 0

        leftMax = [0] * n
        rightMax = [0] * n

        lMax, rMax = height[0], height[n - 1]

        for i in range(1, n - 1):
            leftMax[i] = lMax
            lMax = max(lMax, height[i])

            rightMax[n - i - 1] = rMax
            rMax = max(rMax, height[n - i - 1])
        water = 0
        for i in range(1,n-1) :
            if height[i] < leftMax[i] and height[i] < rightMax[i] :
                water += min(leftMax[i],rightMax[i]) - height[i]
        
        return water

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        if(n < 3) return 0;
        int[] lMaxArr = new int[n];
        int[] rMaxArr = new int[n];
        
        int lMaxVal = height[0]; // Initialize to the first element
        int rMaxVal = height[n - 1]; // Initialize to the last element
        
        for(int i=0;i<n;i++){
            lMaxVal = Math.max(lMaxVal,height[i]);
            lMaxArr[i] = lMaxVal;
            
            rMaxVal = Math.max(rMaxVal,height[n-i-1]);
            rMaxArr[n-i-1] = rMaxVal;
        }
        
        int water = 0;
        
        for(int i=1;i<n-1;i++){
            if(height[i] < lMaxArr[i] && height[i] < rMaxArr[i]){
                water += Math.min(lMaxArr[i],rMaxArr[i]) - height[i];
            }
        }
        
        return water;
    }
}


